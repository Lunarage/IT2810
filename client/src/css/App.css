/* Color codes for the app */
/* Main Primary color */
/* Main Secondary color */
/* Main Tertiary color */
@import url("https://fonts.googleapis.com/css2?family=Merienda:wght@400;700&display=swap");
@import url("https://fonts.googleapis.com/css2?family=Source+Sans+Pro:ital,wght@0,400;0,700;1,400;1,700&display=swap");
body {
  width: 95vw;
  background-color: #e4eedd;
  margin: auto;
  font-family: "Source Sans Pro";
  color: #000000;
}
body * {
  padding: 0;
  margin: 0;
}

.app {
  display: grid;
  grid-template-areas: "banner" "menu" "content";
  grid-template-columns: auto;
  grid-template-rows: auto auto auto;
  grid-gap: 1em;
}

.menu {
  background-color: #cbe2bd;
  grid-area: menu;
  display: grid;
  grid-template-areas: "home search my-page loginbutton";
  grid-template-columns: auto auto auto 1fr;
  justify-content: start;
  padding: 0.3em;
}
.menu a {
  margin: 0.5em 1em;
}
.menu .home-button {
  grid-area: home;
}
.menu .search-button {
  grid-area: search;
}
.menu .my-page-button {
  grid-area: my-page;
}
.menu .login-button {
  justify-self: end;
  grid-area: loginbutton;
}

.content {
  grid-area: content;
  background-color: #F8FEF5;
  justify-self: stretch;
}
.content .start-page {
  align-self: center;
  display: grid;
  justify-content: center;
}

h1 {
  font-weight: bold;
  font-size: 3em;
}

h2 {
  font-weight: bold;
  font-size: 2em;
}

h3 {
  font-size: 1.3em;
  font-weight: bold;
}

a {
  color: #000000;
}
a:hover {
  color: #000000;
  cursor: pointer;
  font-weight: bold;
  text-shadow: 5px 3px 10px #333;
}

@media only screen and (min-width: 600px) {
  body {
    width: 85vw;
  }
}
button {
  border: solid;
  border-color: #c4d3ce;
  border-width: 1px;
  border-radius: 8px;
  background-color: #eaeeed;
  box-shadow: 3px 3px 3px #95b2a9;
  transition: 30ms;
  width: auto;
  padding: 0.5em 1.5em;
  margin: 2px;
  cursor: pointer;
}
button:hover {
  background-color: #c4d3ce;
  border-color: #95b2a9;
  box-shadow: 2px 2px 3px #6b9789;
  transition: 30ms;
}
button:active {
  box-shadow: 0px 0px;
  transition: 30ms;
}
button:focus {
  border-width: 2px;
  border-color: #555;
  outline: none;
  transition: 30ms;
}

input {
  border: solid;
  border-color: #cbe2bd;
  border-width: 2px;
  border-radius: 8px;
  background-color: #FDFFF6;
  transition: 30ms;
  width: auto;
  padding: 0.5em;
  margin: 1px;
}
input:hover {
  box-shadow: 1px 1px 3px #a2c78c;
  transition: 30ms;
}
input:active {
  transition: 15ms;
}
input:focus {
  border-color: #555;
  outline: none;
  transition: 30ms;
}
input:disabled {
  background-color: #dddddd;
}

.login-page {
  display: grid;
  justify-content: center;
}
.login-page .login-section {
  border-radius: 8px;
  background-color: #f6f9e8;
  padding: 2em;
  display: grid;
  grid-template-areas: "title" "form";
  justify-content: center;
  grid-gap: 2em;
}
.login-page .login-section .login-page-title {
  grid-area: title;
  justify-self: center;
}
.login-page .login-section .login-page-form {
  grid-area: form;
  display: grid;
  grid-gap: 1.4em;
  grid-template-areas: "username" "password" "loginbutton";
}

.form-group {
  display: grid;
  grid-template-areas: "label" "input";
}

.search-page {
  display: grid;
  grid-gap: 1em;
  grid-template-areas: "search-bar" "result" "navigation";
  margin: 2em;
}
.search-page .search-bar {
  grid-area: search-bar;
}
.search-page .search-result {
  grid-area: result;
}
.search-page .search-navigation {
  grid-area: navigation;
  justify-self: center;
}
.search-page .search-navigation * {
  margin: 0.5em;
}
.my-page {
  display: grid;
  grid-gap: 1em;
  grid-template-areas: "title" "saved-searches";
  margin: 2em;
}
.my-page .my-page-title {
  justify-self: center;
}
.banner {
  background-color: #a2c78c;
  grid-area: banner;
  display: grid;
  grid-template-areas: "logo banner-login";
  grid-template-rows: auto;
  grid-template-columns: auto 1fr;
}

.logo {
  grid-area: logo;
  font-family: "Merienda";
  font-size: 2em;
  padding: 0.2em;
}

.banner-login {
  grid-area: banner-login;
  align-self: end;
  justify-self: end;
  padding: 0.2em 0.5em;
  color: #808080;
}
.banner-login .login-info {
  display: none;
}
.banner-login .login-info-username {
  font-style: italic;
}

@media only screen and (min-width: 375px) {
  .banner-login {
    display: grid;
    grid-template-areas: "login-info username";
    grid-gap: 0.3em;
  }
  .banner-login .login-info {
    grid-area: login-info;
    display: block;
  }
  .banner-login .login-info-username {
    grid-area: username;
  }
}
@media only screen and (min-width: 600px) and (min-height: 500px) {
  .logo {
    font-size: 4em;
  }
}

/*# sourceMappingURL=App.css.map */
