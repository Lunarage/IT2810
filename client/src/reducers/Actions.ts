import HttpClient from "../modules/HttpClient";
import { SessionStorageWrapper, LocalStorageWrapper } from "../modules/Storage";

export const toggle_loggedIn = (boolVal: boolean) => {
    return {
        type: "TOGGLE_LOGGED_IN",
        payload: boolVal
    } as const
};

export const set_username = ( text: string) => {
    return {
        type: "SET_USERNAME",
        payload: text
    } as const
};

export const logout = (boolVal: boolean) => {
    return {
        type: "LOGOUT",
        payload: boolVal
    } as const 
};
