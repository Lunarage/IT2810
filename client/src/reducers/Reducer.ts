import {
    toggle_loggedIn,
    logout,
    set_username,
    setStoreSearchState,
} from "./Actions";
import { createStore, Store } from "redux";
import { Movie } from "../types/DatabaseTypes";

type Actions =
    | ReturnType<typeof toggle_loggedIn>
    | ReturnType<typeof logout>
    | ReturnType<typeof set_username>
    | ReturnType<typeof setStoreSearchState>;

// The type for holding search information and result
export type StoreSearchState = {
    inputSearch: string;
    inputOrderDir: "ASC" | "DESC";
    inputTitleType: string;
    pageActivePage: number;
    pageTotalPages: number;
    pageDisabled: boolean;
    searchStatus: "none" | "waiting" | "success" | "failure";
    searchErrorMessage: string | null;
    searchMovies: Movie[];
};

//Creates the types AppState
export type AppState = {
    loggedIn: boolean;
    userName?: string | null;
    searchState: StoreSearchState;
};

//Sets initial root state for the app
const initialAppState: AppState = {
    loggedIn: false,
    userName: null,
    searchState: {
        inputSearch: "",
        inputOrderDir: "ASC",
        inputTitleType: "",
        pageActivePage: 1,
        pageTotalPages: 2,
        pageDisabled: false,
        searchStatus: "none",
        searchErrorMessage: null,
        searchMovies: [],
    },
};
//Root reducer for the App, handels the states loggedIn/Loggedout and userName
export const rootReducer = (
    state: AppState = initialAppState,
    action: Actions
) => {
    switch (action.type) {
        case "TOGGLE_LOGGED_IN":
            return {
                ...state,
                loggedIn: action.payload,
            };

        case "SET_USERNAME":
            return {
                ...state,
                userName: action.payload,
            };

        case "LOGOUT":
            return {
                ...state,
                loggedIn: action.payload,
                userName: null,
            };
        case "SET_SEARCH_STATE":
            return {
                ...state,
                searchState: action.payload,
            };
        default:
            return state;
    }
};
//Configure store
function configureStore(): Store<AppState> {
    return createStore(
        rootReducer,
        (window as any).__REDUX_DEVTOOLS_EXTENSION__ &&
            (window as any).__REDUX_DEVTOOLS_EXTENSION__()
    );
}

//Create and exports store
export const store = configureStore();
