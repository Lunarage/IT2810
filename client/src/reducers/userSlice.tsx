import { createSlice } from "@reduxjs/toolkit";
import HttpClient from "../modules/HttpClient";
import { SessionStorageWrapper, LocalStorageWrapper } from "../modules/Storage";

// redux komponent som lager en "slice" av store hvor vi kan definere en reducer og actions som handler om en bruker logget inn eller ikke
const loginSlice = createSlice({
    name: "login/logout",
    initialState: { value: false, username: null },

    reducers: {
        logIn: (state, action) => {
            const baseURL = "http://it2810-22.idi.ntnu.no:3000";
            const client = new HttpClient(baseURL);
            const localStorage = new LocalStorageWrapper();
            localStorage.set("username", action.payload);
            client.createUser(action.payload);
            state.value = true;
            state.username = action.payload;
        },
        logOut: (state) => {
            const localStorage = new LocalStorageWrapper();
            localStorage.remove("username");
            state.value = false;
            state.username = null;
        },
    },
});

export const { logIn, logOut } = loginSlice.actions;

export default loginSlice.reducer;

/* These sites are used as src for this code: 
https://codesandbox.io/s/github/reduxjs/redux-essentials-counter-example/tree/master/?from-embed=&file=/src/features/counter/counterSlice.js:49-83
https://redux.js.org/tutorials/essentials/part-2-app-structure */
