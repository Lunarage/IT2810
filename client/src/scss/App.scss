@import 'colors';

@import 'fonts';

/* Global styling of body */
a {
  color: $color-text-black;

  &:hover {
    color: $color-text-black;
    cursor: pointer;
    font-weight: bold;
    text-decoration: underline;
    //text-shadow: 5px 3px 10px #333; // TODO update shadow color to complement background
  }
}

h1 {
  font-size: 3em;
  font-weight: bold;
}

h2 {
  font-size: 2em;
  font-weight: bold;
}

h3 {
  font-size: 1.3em;
  font-weight: bold;
}

/* Styling page body */
body {
  background-color: $color-primary-0;
  color: $color-text-black;
  font-family: $standard-font;
  margin: auto;
  width: 95vw; // only on small screens

  // Setting padding/margins for all children of body to 0.
  * {
    margin: 0;
    padding: 0;
  }
}

// Styling app
.app {
  // CSS-grid app:
  display: grid;
  grid-gap: 1em;
  grid-template-areas:
    'banner'
    'menu'
    'content';
  grid-template-columns: auto;
  grid-template-rows: auto auto auto;
}

// Styling the menu/nav-bar
.site-menu {
  background-color: $color-primary-3;
  display: grid;
  grid-area: menu;
  grid-template-areas: 'home search my-page loginbutton';
  grid-template-columns: auto auto auto 1fr;
  justify-content: start; // places the inner grid
  padding: 0.3em;

  // Styling menu children (links)
  a,
  p {
    margin: 0.5em 1em;
  }

  p {
    font-weight: bold;
  }

  //Sets grids for buttons in menu
  .home-button {
    grid-area: home;
  }

  .search-button {
    grid-area: search;
  }

  .my-page-button {
    grid-area: my-page;
  }

  // styling menu child login-button
  .login-button {
    grid-area: loginbutton;
    justify-self: end;
  }
}

.content {
  background-color: $color-primary-1;
  grid-area: content;
  justify-self: stretch;
  // Styling the start page
}

@media only screen and (min-width: 600px) {
  body {
    width: 85vw;
  }
}

@import 'button';
@import 'input';
@import 'login-page';
@import 'search-page';
@import 'my-page';
@import 'banner';
@import 'start-page';
