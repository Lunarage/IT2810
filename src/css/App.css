/* Color codes for the app */
/* Main Primary color */
/* Main Secondary color */
/* Main Tertiary color */
@import url("https://fonts.googleapis.com/css2?family=Merienda:wght@400;700&display=swap");
@import url("https://fonts.googleapis.com/css2?family=Source+Sans+Pro:ital,wght@0,400;0,700;1,400;1,700&display=swap");
body {
  width: 95vw;
  background-color: #CBEFB6;
  margin: auto;
  font-family: "Source Sans Pro";
}
body * {
  padding: 0;
  margin: 0;
}

.app {
  display: grid;
  grid-template-areas: "banner" "menu" "content";
  grid-template-columns: auto;
  grid-template-rows: auto auto auto;
  grid-gap: 1em;
}

.menu {
  background-color: #46967C;
  grid-area: menu;
  display: grid;
  grid-template-areas: "button1 button2";
  justify-content: start;
}
.menu a {
  margin: 0.5em 1em;
}

.content {
  grid-area: content;
  background-color: #BFD765;
}
.content .start {
  align-self: center;
  display: grid;
  justify-content: center;
}

h1 {
  font-weight: bold;
  font-size: 3em;
}

h2 {
  font-weight: bold;
  font-size: 2em;
}

h3 {
  font-size: 1.3em;
  font-weight: bold;
}

@media only screen and (min-width: 600px) {
  body {
    width: 85vw;
  }
}
button {
  border: solid;
  border-color: #A2D4C4;
  border-width: 1px;
  border-radius: 8px;
  background-color: #D0ECE3;
  box-shadow: 3px 3px 5px #71B49E;
  transition: 30ms;
  width: auto;
  padding: 0.5em 1.5em;
  cursor: pointer;
}
button:hover {
  background-color: #A2D4C4;
  border-color: #71B49E;
  box-shadow: 2px 2px 4px #46967C;
  transition: 30ms;
}
button:active {
  box-shadow: 0px 0px;
  transition: 30ms;
}
button:focus {
  border-color: #555;
  outline: none;
  transition: 30ms;
}

input {
  border: solid;
  border-color: #ADE38E;
  border-width: 2px;
  border-radius: 8px;
  background-color: #FDFFF6;
  transition: 30ms;
  width: auto;
  padding: 0.5em;
  margin: 1px;
  cursor: pointer;
}
input:hover {
  box-shadow: 1px 1px 3px #83C55C;
  transition: 30ms;
}
input:active {
  transition: 15ms;
}
input:focus {
  border-color: #555;
  outline: none;
  transition: 30ms;
}

.banner {
  background-color: #83C55C;
  grid-area: banner;
  display: grid;
  grid-template-areas: "logo banner-login";
  grid-template-rows: auto;
  grid-template-columns: auto 1fr;
}

.logo {
  grid-area: logo;
  font-family: "Merienda";
  font-size: 4em;
}

.banner-login {
  grid-area: banner-login;
  align-self: start;
  justify-self: end;
  display: grid;
  grid-template-areas: "login-info login-button";
  grid-gap: 1em;
}
.banner-login .login-info {
  grid-area: login-info;
}
.banner-login .login-button {
  grid-area: login-button;
}

/*# sourceMappingURL=App.css.map */
