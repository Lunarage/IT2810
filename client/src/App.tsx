import React from "react";
import logo from "./logo.svg";
import "./css/App.css";
import Banner from "./components/Banner";
import SearchPage from "./components/SearchPage";
import LoginPage from "./components/LoginPage";
import MyPage from "./components/MyPage";
import Menu from "./components/Menu";
import StartPage from "./components/StartPage";

class App extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      homePage: true,
      searchPage: false,
      myPage: false,
      loginPage: false,
    };
    this.handleHomePageClick = this.handleHomePageClick.bind(this);
    this.handleMyPageClick = this.handleMyPageClick.bind(this);
    this.handleSearchPageClick = this.handleSearchPageClick.bind(this);
    this.handleLoginPageClick = this.handleLoginPageClick.bind(this);
    this.handleLogoClick = this.handleLogoClick.bind(this);
  }
  /*These functions toggle on click the state of the App, so the page shown changes when the 
  buttons in the menu are pressed, they are passed ass props to the menu button*/

  handleMyPageClick() {
    this.setState({
      homePage: false,
      searchPage: false,
      myPage: true,
      loginPage: false,
    });
  }

  handleSearchPageClick() {
    this.setState({
      homePage: false,
      searchPage: true,
      myPage: false,
      loginPage: false,
    });
  }

  handleHomePageClick() {
    this.setState({
      homePage: true,
      searchPage: false,
      myPage: false,
      loginPage: false,
    });
  }

  handleLoginPageClick() {
    this.setState({
      homePage: false,
      searchPage: false,
      myPage: false,
      loginPage: true,
    });
  }
  
  handleLogoClick() {
    this.setState({
      homePage: true,
      searchPage: false,
      myPage: false,
      loginPage: false,
    });
  }

  render() {
    //The renderPage() function checks the state of App and based on that what should be rendered
    const renderPage = () => {
      if (
        this.state.homePage === false &&
        this.state.searchPage === false &&
        this.state.myPage === true &&
        this.state.loginPage === false
      ) {
        return <MyPage />;
      }
      if (
        this.state.homePage === true &&
        this.state.searchPage === false &&
        this.state.myPage === false &&
        this.state.loginPage === false
      ) {
        return <StartPage />;
      }
      if (
        this.state.homePage === false &&
        this.state.searchPage === true &&
        this.state.myPage === false &&
        this.state.loginPage === false
      ) {
        return <SearchPage />;
      }
      if (
        this.state.homePage === false &&
        this.state.searchPage === false &&
        this.state.myPage === false &&
        this.state.loginPage === true
      ) {
        return <LoginPage />;
      }
    };
    return (
      <div className="app">
        <Banner onLogoClick={this.handleLogoClick} />
        <Menu
          onMyPageClick={this.handleMyPageClick}
          onHomePageClick={this.handleHomePageClick}
          onSearchPageClick={this.handleSearchPageClick}
          onLoginPageClick={this.handleLoginPageClick}
        />
        <div className={"content"}>{renderPage()}</div>
      </div>
    );
  }
}

export default App;

interface Props {}

interface State {
  homePage: boolean;
  searchPage: boolean;
  myPage: boolean;
  loginPage: boolean;
}
