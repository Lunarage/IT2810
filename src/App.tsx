import React from 'react';
import logo from './logo.svg';
import './css/App.css';
import Banner from "./components/Banner";
import SearchPage from "./components/SearchPage";
import LoginPage from "./components/LoginPage";
import MyPage from "./components/MyPage";

function App() {
    return (
        <div className="App">
            <header className="App-header">
                <Banner/>
            </header>
            <body>
                <SearchPage/>
                <LoginPage/>
                <MyPage/>
            </body>
        </div>
    );
}

export default App;
