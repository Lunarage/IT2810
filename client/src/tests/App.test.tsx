import App from "../App"
import { Provider } from "react-redux"
import { store } from "../reducers/Reducer"
import React from "react";
import { render } from "@testing-library/react";

it('App renders', () => {
  const {container} = render(
    <Provider store={store}><App/></Provider>);
  expect(container).toMatchSnapshot();
});