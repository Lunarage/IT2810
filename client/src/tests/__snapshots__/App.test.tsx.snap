// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<div>
  <div
    class="app"
  >
    <div
      class="banner"
    >
      <h1
        class="logo"
      >
        Logo
      </h1>
    </div>
    <div
      class="site-menu"
    >
      <p>
        Home
      </p>
      <a
        class="search-button"
        tabindex="0"
      >
        Search
      </a>
      <button
        class="login-button"
      >
        Log in
      </button>
    </div>
    <div
      class="content"
    >
      <div
        class="start-page"
      >
        <h2
          class="welcome"
        >
          Welcome!
        </h2>
        <p
          class="welcome-p"
        >
          This website gives you the oppurtunity to search in over 7.2 million movies, TV-series and other film media
        </p>
        <h3
          class="how-to-use"
        >
          How to use?
        </h3>
        <p
          class="how-to-use-p"
        >
          In the search section of webpage you can type in any word or title of a movie and will get all the matches from the database. you can also like your favorite movies if you are logged in. When you are logged in, myPage will appear on the site and here you can see all your liked movies and shows.
        </p>
        <p
          class="how-to-use-p"
        >
          To log in you can either write user1 in the username field and see an example of a user with data, or you can write any other name in the user field and make your own data. Password is not required.
        </p>
      </div>
    </div>
  </div>
</div>
`;
